### PR TITLE
Conditional string type conversion

### PR DESCRIPTION
**Pull request by saschanaz**
`.encode('utf-8')` is required to convert Unicode string to Python 2 native byte string type, while the encoding must not occur in Python 3 as it causes type mismatch where the native type is Unicode. This PR thus encode strings only when the native type is `bytes`.

### CODE REVIEW

**Comment by kripken** [comment]
This is a little verbose, and we'd need it in more places too, won't we? Perhaps we could have a utility function that does this, so it's just a call instead of several lines in each location? I wonder what other projects do for this.

**Comment by saschanaz** [comment]
Numpy has [a utility function](https://github.com/numpy/numpy/blob/5d914da9846159c1863f926bedacb97b08f8bda3/tools/npy_tempita/compat3.py#L23) named `b`:

```py
if sys.version_info[0] < 3:

    def next(obj):
        return obj.next()

    def iteritems(d, **kw):
        return d.iteritems(**kw)

    b = bytes = str
    basestring_ = basestring

else:

    def b(s):
        if isinstance(s, str):
            return s.encode('latin1')
        return bytes(s)

    def iteritems(d, **kw):
        return iter(d.items(**kw))

    next = next
    basestring_ = (bytes, str)
    bytes = bytes

text = str
```

**Comment by saschanaz** [comment]
Numpy has [`asstr` helper function](https://github.com/numpy/numpy/blob/e5d7653f8456c66ccb005eb50a139b1a9b044b3f/numpy/compat/py3k.py#L37):

```py
def asstr(s):
    if isinstance(s, bytes):
        return s.decode('latin1')
    return str(s)
```

...which will cause double conversion on Py2. Not sure why they do in that way but it anyway will work.

**Comment by kripken** [comment]
Double conversion worries me because we do have some very large strings (the entire compiler output), but I'm not sure if those would go through that code path.

**Comment by saschanaz** [comment]
Ah, I misread that function. The helper function always returns Unicode string for Python 3, and an in Python 2 it is defined as an alias of `str`. (I think I need some more coffee this week...)

I think we can do similar thing:

```py
def asstr(s):
  if str is bytes:
    if isinstance(s, unicode):
      return s.encode('utf-8')
  elif isinstance(s, bytes):
      return s.decode('utf-8')
```

File: emscripten.py [hunk 1]
```diff
@@ -2152,9 +2152,7 @@ def main(args, compiler_engine, cache, temp_files, DEBUG):
   for setting in args.settings:
     name, value = setting.strip().split('=', 1)
     value = json.loads(value)
-    if isinstance(value, unicode):
```
Comment (kripken) [inline emscripten.py]: can't we use `asstr` in these places?

I admit I am getting confused with `asstr, asbytes` and others. Perhaps we should write up a Python2/3 developer's guide, explaining the proper best practices?

Comment (saschanaz) [reply emscripten.py]: >can't we use asstr in these places?

We can, thus the line 2155?

Comment (kripken) [reply emscripten.py]: Oh sorry, I completely misread this diff... :)

File: tests/runner.py [hunk 2]
```diff
@@ -521,7 +521,10 @@ def assertTextDataContained(self, text1, text2):
   def assertContained(self, values, string, additional_info=''):
     if type(values) not in [list, tuple]: values = [values]
     for value in values:
-      if not isinstance(value, bytes): string = string.decode('UTF-8') # If we have any non-ASCII chars in the expected string, treat the test string from ASCII as UTF8 as well.
+      if str is bytes and type(value) is unicode:
```
Comment (kripken) [inline tests/runner.py]: is it worth making a helper function for this case too?

File: tests/test_core.py [hunk 3]
```diff
@@ -6680,7 +6680,10 @@ def encode_utf8(data):
         else:
           return data
 
-      data = encode_utf8(json.load(open(map_filename, 'r')))
+      data = json.load(open(map_filename, 'r'))
+      if str is bytes:
```
Comment (kripken) [inline tests/test_core.py]: could this be `asstr`?

Comment (saschanaz) [reply tests/test_core.py]: `data` is not a string but a dictionary so it can't.

File: tests/runner.py (hunk 2 above)
Comment (saschanaz) [reply tests/runner.py]: Slightly wired logic here, we may do values = map(asstr, values) instead.

**Comment by saschanaz** [comment]
I think now it's safe to merge, what do you think?

File: tests/test_core.py (hunk 3 above)
Comment (kripken) [reply tests/test_core.py]: oh, right, thanks.

**Pull request merged**